### PR TITLE
Implement SDL-0116 Open Menu RPC

### LINKED ISSUE #806: [SDL 0116] Open Menu RPC
Proposal: [Open Menu RPC](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0116-open-menu.md)

> This proposal is about providing the capability to open the built-in app-menu which is provided by the native UI.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/352

Steering Committee Decision:

> The Steering Committee voted to accept this proposal, with the revision outlined in kshala-ford's last [comment](https://github.com/smartdevicelink/sdl_evolution/issues/352#issuecomment-347462343). There was also discussion about allowing an RPC to close the menu, but it was determined that this proposal works to address when there isn't a system button on the screen, and there would always be a system button giving the user the option to close.

>The proposal .md was updated on December 4, 2017 to include the agreed upon revisions.

### PR DESCRIPTION
**Pull request by justingluck93**
Fixes #806 

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
Send a RPC request to HMI to open the menu
```
SDLShowAppMenu *showMenu = [[SDLShowAppMenu alloc] init];
[self.sdlManager sendRequest:showMenu withResponseHandler:^(__kindof SDLRPCRequest * _Nullable request, __kindof SDLRPCResponse * _Nullable response, NSError * _Nullable error) {
       if (error != nil) { return; }
       // Use Response if needed
    }];
```

```
SDLShowAppMenu *showMenu = [[SDLShowAppMenu alloc] init];
showMenu.menuID = //MenuID With SubCells
[self.sdlManager sendRequest:showMenu withResponseHandler:^(__kindof SDLRPCRequest * _Nullable request, __kindof SDLRPCResponse * _Nullable response, NSError * _Nullable error) {
       if (error != nil) { return; }
       // Use Response if needed
    }];
```

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: SmartDeviceLink/SDLShowAppMenu.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+NS_ASSUME_NONNULL_BEGIN
+
+/**
+ *  Used by an app typically of navigation type to show the apps menu.
+ */
+@interface SDLShowAppMenu : SDLRPCRequest
+
+/**
+ * a Menu ID that identifies the sub menu to open. If not set the top level menu will be opened.
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLShowAppMenu.h]:
```suggestion
 A Menu ID that identifies the sub menu to open. If not set the top level menu will be opened.
```

File: SmartDeviceLink/SDLShowAppMenu.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+//
+
+#import "SDLRPCRequest.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+/**
+ *  Used by an app typically of navigation type to show the apps menu.
+ */
+@interface SDLShowAppMenu : SDLRPCRequest
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLShowAppMenu.h]: Add an `initWithMenuID:` initializer

File: SmartDeviceLinkTests/RPCSpecs/RequestSpecs/SDLShowAppMenuSpec.m [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    it(@"Should set and get correctly", ^ {
+        SDLShowAppMenu *testRequest = [[SDLShowAppMenu alloc] init];
+
+        testRequest.menuID = @(menuId);
+
+        expect(testRequest.menuID).to(equal(testRequest.menuID));
+    });
+
+    it(@"Should get correctly when initialized with dictonary", ^ {
```
Comment (joeljfischer) [inline SmartDeviceLinkTests/RPCSpecs/RequestSpecs/SDLShowAppMenuSpec.m]: Shouldn't use mutable dictionaries

File: SmartDeviceLink/SDLShowAppMenu.h (hunk 2 above)
Comment (justingluck93) [reply SmartDeviceLink/SDLShowAppMenu.h]: added

File: SmartDeviceLinkTests/RPCSpecs/RequestSpecs/SDLShowAppMenuSpec.m (hunk 3 above)
Comment (justingluck93) [reply SmartDeviceLinkTests/RPCSpecs/RequestSpecs/SDLShowAppMenuSpec.m]: fixed

File: SmartDeviceLink/SDLScreenManager.h [hunk 4]
```diff
@@ -277,6 +277,11 @@ If set to `SDLDynamicMenuUpdatesModeForceOff`, menu updates will work the legacy
  */
 - (void)presentKeyboardWithInitialText:(NSString *)initialText delegate:(id<SDLKeyboardDelegate>)delegate;
 
+#pragma mark Menu
+
+- (BOOL)openMenu;
+- (BOOL)openSubmenu:(SDLMenuCell *)cell;
```
[the remark this answers is not in the document]
Comment (justingluck93) [reply SmartDeviceLink/SDLScreenManager.h]: added

File: SmartDeviceLink/SDLScreenManager.m [hunk 5]
```diff
@@ -264,6 +265,25 @@ - (void)presentKeyboardWithInitialText:(NSString *)initialText delegate:(id<SDLK
...
+- (BOOL)openMenu {
+    if([SDLGlobals.sharedGlobals.rpcVersion isLessThanVersion:[[SDLVersion alloc] initWithMajor:6 minor:0 patch:0]]) {
+        return NO;
+    }
+    [self.menuManager openMenu];
+    return YES;
+}
+
+- (BOOL)openSubmenu:(SDLMenuCell *)cell {
+    if([SDLGlobals.sharedGlobals.rpcVersion isLessThanVersion:[[SDLVersion alloc] initWithMajor:6 minor:0 patch:0]]) {
```
[the remark this answers is not in the document]
Comment (justingluck93) [reply SmartDeviceLink/SDLScreenManager.m]: added

File: SmartDeviceLink/SDLScreenManager.m [hunk 6]
```diff
@@ -264,6 +265,25 @@ - (void)presentKeyboardWithInitialText:(NSString *)initialText delegate:(id<SDLK
     [self.choiceSetManager presentKeyboardWithInitialText:initialText delegate:delegate];
 }
 
+#pragma mark - Menu
+
+- (BOOL)openMenu {
+    if([SDLGlobals.sharedGlobals.rpcVersion isLessThanVersion:[[SDLVersion alloc] initWithMajor:6 minor:0 patch:0]]) {
```
[the remark this answers is not in the document]
Comment (justingluck93) [reply SmartDeviceLink/SDLScreenManager.m]: added

File: SmartDeviceLink/SDLShowAppMenu.h [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ *  Convenience init for setting menuID
+ *
+ *  @param menuID       Menu id of requsted sub menu
+ *
+ *  @return             A SDLShowAppMenu object
```
[the remark this answers is not in the document]
Comment (justingluck93) [reply SmartDeviceLink/SDLShowAppMenu.h]: added

File: SmartDeviceLink/SDLMenuManager.m [hunk 8]
```diff
@@ -638,6 +639,30 @@ - (void)sdl_hmiStatusNotification:(SDLRPCNotificationNotification *)notification
...
+    }];
+}
+
+- (void)openSubmenu:(SDLMenuCell *)cell {
+    if (cell.subCells == 0) {
+        SDLLogW(@"The cell does not contain any sub cells, RPC will not be sent");
+        return;
+    }
+
+    SDLShowAppMenu *subMenu = [[SDLShowAppMenu alloc] initWithMenuID:cell.cellId];
```
[the remark this answers is not in the document]
Comment (justingluck93) [reply SmartDeviceLink/SDLMenuManager.m]: added a check to see if cell is in the array

File: SmartDeviceLink/SDLMenuManager.h [hunk 9]
```diff
@@ -36,6 +36,10 @@ typedef void(^SDLMenuUpdateCompletionHandler)(NSError *__nullable error);
 @property (copy, nonatomic) NSArray<SDLMenuCell *> *menuCells;
 
 @property (assign, nonatomic) SDLDynamicMenuUpdatesMode dynamicMenuUpdatesMode;
+
+- (void)openMenu;
+- (void)openSubmenu:(SDLMenuCell *)cell;
```
[the remark this answers is not in the document]
Comment (justingluck93) [reply SmartDeviceLink/SDLMenuManager.h]: added

File: SmartDeviceLink/SDLMenuManager.m [hunk 10]
```diff
@@ -638,6 +639,30 @@ - (void)sdl_hmiStatusNotification:(SDLRPCNotificationNotification *)notification
...
+
+    [self.connectionManager sendConnectionRequest:subMenu withResponseHandler:^(__kindof SDLRPCRequest * _Nullable request, __kindof SDLRPCResponse * _Nullable response, NSError * _Nullable error) {
+        if (error != nil) {
+            SDLLogE(@"Error opening application menu: %@", error);
```
[the remark this answers is not in the document]
Comment (justingluck93) [reply SmartDeviceLink/SDLMenuManager.m]: fixed, chose to implement myself to keep all PR issues to one commit

File: SmartDeviceLink/SDLMenuManager.m [hunk 11]
```diff
@@ -638,6 +639,30 @@ - (void)sdl_hmiStatusNotification:(SDLRPCNotificationNotification *)notification
...
+            return;
+        }
+    }];
+}
+
+- (void)openSubmenu:(SDLMenuCell *)cell {
+    if (cell.subCells == 0) {
+        SDLLogW(@"The cell does not contain any sub cells, RPC will not be sent");
```
[the remark this answers is not in the document]
Comment (justingluck93) [reply SmartDeviceLink/SDLMenuManager.m]: fixed, chose to implement myself to keep all PR issues to one commit

File: SmartDeviceLink/SDLMenuManager.m [hunk 12]
```diff
@@ -638,6 +641,46 @@ - (void)sdl_hmiStatusNotification:(SDLRPCNotificationNotification *)notification
     }
 }
 
+- (BOOL)openMenu {
+    if([SDLGlobals.sharedGlobals.rpcVersion isLessThanVersion:[[SDLVersion alloc] initWithMajor:6 minor:0 patch:0]]) {
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLMenuManager.m]:
```suggestion
    if ([SDLGlobals.sharedGlobals.rpcVersion isLessThanVersion:[[SDLVersion alloc] initWithMajor:6 minor:0 patch:0]]) {
```

File: SmartDeviceLink/SDLMenuManager.m [hunk 13]
```diff
@@ -638,6 +641,46 @@ - (void)sdl_hmiStatusNotification:(SDLRPCNotificationNotification *)notification
...
+    }];
+
+    return YES;
+}
+
+- (BOOL)openSubmenu:(SDLMenuCell *)cell {
+    if (cell.subCells.count == 0) {
+        SDLLogE(@"The cell %@ does not contain any sub cells, so no submenu can be opened", cell);
+        return NO;
+    } else if([SDLGlobals.sharedGlobals.rpcVersion isLessThanVersion:[[SDLVersion alloc] initWithMajor:6 minor:0 patch:0]]) {
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLMenuManager.m]:
```suggestion
    } else if ([SDLGlobals.sharedGlobals.rpcVersion isLessThanVersion:[[SDLVersion alloc] initWithMajor:6 minor:0 patch:0]]) {
```

File: SmartDeviceLink/SDLMenuManager.m [hunk 14]
```diff
@@ -638,6 +641,46 @@ - (void)sdl_hmiStatusNotification:(SDLRPCNotificationNotification *)notification
...
+        SDLLogE(@"The openSubmenu method is not supported on this head unit.");
+        return NO;
+    } else if(![self.menuCells containsObject:cell]) {
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLMenuManager.m]:
```suggestion
    } else if (![self.menuCells containsObject:cell]) {
```

**Pull request merged**